File: app/schemas/recipe.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime

class RecipeBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200, description="Recipe name")
    description: Optional[str] = Field(None, max_length=1000, description="Recipe description")
    difficulty: str = Field(..., description="난이도 (ex: 쉬움, 보통, 어려움)")
    cooking_time: int = Field(..., ge=1, description="Cooking time in minutes")
    servings: int = Field(..., ge=1, description="Number of servings")
    instructions: str = Field(..., min_length=1, description="Cooking instructions")
    image_url: Optional[str] = Field(None, description="Recipe image URL")

class RecipeCreate(RecipeBase):
    pass

class RecipeUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    difficulty: Optional[str] = Field(None, description="난이도 (ex: 쉬움, 보통, 어려움)")
    cooking_time: Optional[int] = Field(None, ge=1)
    servings: Optional[int] = Field(None, ge=1)
    instructions: Optional[str] = Field(None, min_length=1)
    image_url: Optional[str] = None

class RecipeResponse(RecipeBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class RecipeListResponse(BaseModel):
    recipes: List[RecipeResponse]
    total: int
    page: int
    size: int

# 레시피 검색 관련 스키마
class RecipeIngredient(BaseModel):
    name: str
    amount: Optional[float] = None
    unit: Optional[str] = None
    price: Optional[float] = None

class RecipeInstruction(BaseModel):
    step: str
    number: Optional[int] = None

class RecipeNutrition(BaseModel):
    calories: Optional[float] = None
    protein: Optional[float] = None
    fat: Optional[float] = None
    carbohydrates: Optional[float] = None
    fiber: Optional[float] = None

class RecipeRecommendation(BaseModel):
    id: Optional[int] = None
    title: str
    summary: Optional[str] = None
    image_url: Optional[str] = None
    ingredients: List[RecipeIngredient] = []
    instructions: List[RecipeInstruction] = []
    nutrition: Optional[RecipeNutrition] = None
    cooking_time: Optional[int] = None
    servings: Optional[int] = None
    difficulty: Optional[str] = None
    source: str
    total_cost: Optional[float] = None
    currency: str = "KRW"

class RecipeSearchRequest(BaseModel):
    query: str = Field(..., description="검색할 레시피 이름")
    number: int = Field(3, ge=1, le=50, description="반환할 레시피 개수")
    include_price: bool = Field(False, description="가격 정보 포함 여부")
    max_cooking_time: Optional[int] = Field(None, ge=1, description="최대 조리 시간 (분)")
    cuisine_type: Optional[str] = Field(None, description="요리 타입")

class RecipeSearchResponse(BaseModel):
    query: str
    total_results: int
    recipes: List[RecipeRecommendation]
    estimated_total_cost: Optional[float] = None
    currency: str = "KRW" 

File: frontend/src/types/recipe.ts
export interface RecipeIngredient {
  name: string;
  amount?: string;
  unit?: string;
  price?: number;
}

export interface RecipeInstruction {
  step: string;
  number?: number;
}

export interface AnalyzedInstructionStep {
  number: number;
  step: string;
  ingredients?: Array<{
    id: number;
    name: string;
    localizedName?: string;
    image?: string;
  }>;
  equipment?: Array<{
    id: number;
    name: string;
    localizedName?: string;
    image?: string;
  }>;
  length?: {
    number: number;
    unit: string;
  };
}

export interface AnalyzedInstructionGroup {
  name?: string;
  steps: AnalyzedInstructionStep[];
}

export interface RecipeNutrition {
  calories?: number;
  protein?: number;
  fat?: number;
  carbohydrates?: number;
  fiber?: number;
}

export interface RecipeRecommendation {
  id?: number;
  title: string;
  summary?: string;
  image_url?: string;
  image?: string;
  readyInMinutes?: number;
  cuisines?: string[];
  dishTypes?: string[];
  diets?: string[];
  extendedIngredients?: ExtendedIngredient[];
  sourceName?: string;
  sourceUrl?: string;
  ingredients?: RecipeIngredient[];
  instructions?: RecipeInstruction[] | string;
  analyzedInstructions?: AnalyzedInstructionGroup[];
  nutrition?: RecipeNutrition;
  cooking_time?: number;
  servings?: number;
  difficulty?: string;
  source: string;
  total_cost?: number;
  currency: string;
  pricePerServing?: number;
  aggregateLikes?: number;
  vegetarian?: boolean;
  vegan?: boolean;
  glutenFree?: boolean;
  dairyFree?: boolean;
  veryHealthy?: boolean;
  cheap?: boolean;
  [key: string]: any;
}

export interface ExtendedIngredient {
  id: number;
  aisle?: string;
  image?: string;
  consistency?: string;
  name: string;
  nameClean?: string;
  original: string;
  originalName?: string;
  amount: number;
  unit: string;
  meta?: string[];
  measures?: {
    us?: {
      amount: number;
      unitShort: string;
      unitLong: string;
    };
    metric?: {
      amount: number;
      unitShort: string;
      unitLong: string;
    };
  };
}

export interface RecipeSearchRequest {
  query: string;
  number: number;
  include_price: boolean;
  max_cooking_time?: number;
  cuisine_type?: string;
}

export interface RecipeSearchResponse {
  query: string;
  total_results: number;
  recipes: RecipeRecommendation[];
  estimated_total_cost?: number;
  currency: string;
}

const tagMap: Record<string, string> = {
  // cuisines
  Chinese: '중식',
  Asian: '아시아',
  // dishTypes
  'main course': '메인 코스',
  'side dish': '사이드',
  lunch: '점심',
  dinner: '저녁',
  // diets
  'gluten free': '글루텐 프리',
  'dairy free': '유제품 프리',
  vegetarian: '채식',
  vegan: '비건',
  // ...필요한 만큼 추가
};

function translateTag(tag: string) {
  return tagMap[tag] || tag;
} 